File: 2058.cpp
struct Solution {
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
        int firstCrit = -1, lastCrit = -1, minDist = 100001, i = 1, prevVal = head->val;
        ListNode *ptr = head->next, *nextPtr = head->next->next;
        while (nextPtr != nullptr){
            if ((ptr->val > prevVal && ptr->val > nextPtr->val) || (ptr->val < prevVal && ptr->val < nextPtr->val)){
                if (firstCrit == -1){
                    firstCrit = i;
                }
                if (lastCrit != -1){
                    minDist = min(minDist, i - lastCrit);
                }
                lastCrit = i;
            }
            prevVal = ptr->val;
            ++i;
            ptr = nextPtr;
            nextPtr = nextPtr->next;
        }
        if (firstCrit == -1 || firstCrit == lastCrit){
            return {-1, -1};
        } else {
            return {minDist, lastCrit - firstCrit};
        }
    }
};

File: sol/2058/2058.py
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        firstCrit = None
        lastCrit = None
        minDist = int(10 ** 5) + 1
        ptr = head.next
        prevVal = head.val
        i = 1
        while ptr.next is not None:
            if (ptr.val > prevVal and ptr.val > ptr.next.val) or (ptr.val < prevVal and ptr.val < ptr.next.val):
                if firstCrit is None:
                    firstCrit = i
                if lastCrit is not None:
                    minDist = min(minDist, i - lastCrit)
                lastCrit = i
            prevVal = ptr.val
            i += 1
            ptr = ptr.next
        if firstCrit is None or firstCrit == lastCrit:
            return [-1, -1]
        else:
            return [minDist, lastCrit - firstCrit]
